### ISSUE TITLE
-pie should be a bin_ldflags, not bin_cflags

### ISSUE THREAD

**Issue by noloader** [openssl/openssl#11237]
I'm building OpenSSL for Android for API 23 using NDK-r20. Android needs `-pie` for executables for the last several years. The requirement took effect around Android Lollipop from circa 2014.

Currently [`15-android.conf`](https://github.com/openssl/openssl/blob/master/Configurations/15-android.conf#L153) specifies `-pie` for `bin_cflags` (comments removed):

```
    "android" => {
        inherit_from     => [ "linux-generic32" ],
        template         => 1,
        cflags           => add(sub { android_ndk()->{cflags} }),
        cppflags         => add(sub { android_ndk()->{cppflags} }),
        cxxflags         => add(sub { android_ndk()->{cflags} }),
        bn_ops           => sub { android_ndk()->{bn_ops} },
        bin_cflags       => "-pie",
        enable           => [ ],
    },
```

Using `-pie` for `CFLAGS` results in a [warning for each file compiled](https://travis-ci.org/NLnetLabs/unbound/jobs/657892115):

```
armv7a-linux-androideabi23-clang  -I. -Iinclude -Iapps -pie -pthread -D__ANDROID_API__=23 -march=armv7-a -mthumb -mfloat-
abi=softfp -funwind-tables -fexceptions --sysroot=/home/travis/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot
-D__ANDROID_API__=23 -DNDEBUG -D__ANDROID_API__=23 -march=armv7-a -mthumb -mfloat-abi=softfp -funwind-tables -fexceptions
--sysroot=/home/travis/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot -MMD -MF apps/asn1pars.d.tmp -MT
apps/asn1pars.o -c -o apps/asn1pars.o apps/asn1pars.c

clang: warning: argument unused during compilation: '-pie' [-Wunused-command-line-argument]
```

The problem is, there is no `bin_ldflags`, so we can't change to a more appropriate setting. That is, we can't say `bin_ldflags => "-pie"`:

```
$ git branch
* OpenSSL_1_1_1-stable
$ grep -wIR bin_ldflags ./*
$ 
```

**Comment by kroeckx** [comment]
Don't the CFLAGS also need -fPIE? But the compiler probably does
-fPIE by default.

**Comment by levitte** [comment]
The name is `bin_lflags`.  I know, confusing... historical reasons.

I was about to ask exactly the same as @kroeckx did there

**Comment by levitte** [comment]
I can't see -fPIE in the clang manual, so I assume it's simply ignored, and that clang does the right thing.  However, if gcc is used, I assume we will indeed need -fPIE.  It would be nice if this could be confirmed...

**Comment by levitte** [comment]
@noloader, would you mind trying this quick patch?

``` diff
diff --git a/Configurations/15-android.conf b/Configurations/15-android.conf
index b28689a425..7d5a503c16 100644
--- a/Configurations/15-android.conf
+++ b/Configurations/15-android.conf
@@ -168,7 +168,8 @@ my %targets = (
         cppflags         => add(sub { android_ndk()->{cppflags} }),
         cxxflags         => add(sub { android_ndk()->{cflags} }),
         bn_ops           => sub { android_ndk()->{bn_ops} },
-        bin_cflags       => "-pie",
+        bin_cflags       => "-fPIE",
+        bin_lflags       => "-pie",
         enable           => [ ],
     },
     "android-arm" => {
```

**Comment by noloader** [comment]
@levitte,

So I am clear... You want this tested, correct?

```
    "android" => {
        inherit_from     => [ "linux-generic32" ],
        template         => 1,
        ################################################################
        # Special note about -pie. The underlying reason is that
        # Lollipop refuses to run non-PIE. But what about older systems
        # and NDKs? -fPIC was never problem, so the only concern is -pie.
        cflags           => add(sub { android_ndk()->{cflags} }),
        cppflags         => add(sub { android_ndk()->{cppflags} }),
        cxxflags         => add(sub { android_ndk()->{cflags} }),
        bn_ops           => sub { android_ndk()->{bn_ops} },
        bin_cflags       => "-fPIE",
        bin_lflags       => "-pie",
        enable           => [ ],
    },
```

**Comment by noloader** [comment]
The job is running on Travis at [Unbound | 657947227](https://travis-ci.org/noloader/unbound/builds/657947227).

**Comment by levitte** [comment]
Okie, fix in #11238
